WITH_LIRIC: implicit_interface_29 link fails with duplicate .str.* / ERROR STOP symbols

## Summary
`WITH_LIRIC` fails to link `implicit_interface_29` in LFortran integration tests due to duplicate global symbols (`.str.*`, `ERROR STOP`).

This is a **Liric-side parity issue** because the same targeted integration test passes with LLVM build and fails only with WITH_LIRIC build.

## Current behavior
Link step fails with many errors like:
- `multiple definition of '.str.1'`
- `multiple definition of 'ERROR STOP'`

Log evidence:
- `/tmp/liric_lfortran_api_compat_out/logs/lfortran_integration_tests_liric_api.log`

## Reproduction
### LLVM build (passes)
```bash
cd /home/ert/code/lfortran-dev/lfortran/integration_tests
PATH=/home/ert/code/lfortran-dev/lfortran/build/src/bin:$PATH \
  python run_tests.py -b llvm --ninja -j4 -t implicit_interface_29
```

### WITH_LIRIC build (fails)
```bash
cd /home/ert/code/lfortran-dev/lfortran/integration_tests
PATH=/home/ert/code/lfortran-dev/lfortran/build-liric/src/bin:$PATH \
LFORTRAN_LINKER=gcc \
  python run_tests.py -b llvm --ninja -j4 -t implicit_interface_29
```

## Deep evidence
Failing object files contain exported duplicate data symbols:
```bash
nm -A /tmp/liric_lfortran_api_compat/lfortran/integration_tests/test-llvm/CMakeFiles/implicit_interface_29.dir/implicit_interface_29.f90.o | rg '\.str\.|ERROR STOP'
nm -A /tmp/liric_lfortran_api_compat/lfortran/integration_tests/test-llvm/CMakeFiles/implicit_interface_29.dir/implicit_interface_29b.f90.o | rg '\.str\.|ERROR STOP'
```

## Proposed fix scope (Liric)
1. Make compiler-generated string/error-stop constants non-exported across translation units (LLVM-equivalent private/internal linkage semantics).
2. Ensure symbol visibility/linkage is propagated from LLVM-compat APIs into Liric IR/object output.
3. Add regression to guarantee two units with internal constants never collide at link.

## Verification / acceptance criteria
- Targeted case passes in WITH_LIRIC:
```bash
cd /home/ert/code/lfortran-dev/lfortran/integration_tests
PATH=/home/ert/code/lfortran-dev/lfortran/build-liric/src/bin:$PATH \
LFORTRAN_LINKER=gcc \
  python run_tests.py -b llvm --ninja -j4 -t implicit_interface_29
```
- Full compat harness still run:
```bash
cd /home/ert/code/lfortran-dev/liric
./tools/lfortran_mass/lfortran_api_compat.sh --fresh-workspace
```

--- a/include/llvm/IR/Function.h
+++ b/include/llvm/IR/Function.h
@@ -82,6 +82,9 @@ private:
 
 public:
     Function() : func_val_(nullptr), compat_mod_(nullptr), block_list_(this) {}
+    ~Function() {
+        detail::unregister_global_value_state(this);
+    }
 
     void setFuncVal(lc_value_t *fv) { func_val_ = fv; }
     lc_value_t *getFuncVal() const { return func_val_; }

--- a/include/llvm/IR/GlobalValue.h
+++ b/include/llvm/IR/GlobalValue.h
@@ -2,11 +2,43 @@
 #define LLVM_IR_GLOBALVALUE_H
 
 #include "llvm/IR/Constants.h"
+#include <unordered_map>
 
 namespace llvm {
 
 class Module;
 class PointerType;
+
+namespace detail {
+    struct global_value_state {
+        int linkage = 0;
+        int visibility = 0;
+        int unnamed_addr = 0;
+    };
+
+    inline thread_local std::unordered_map<const void *, global_value_state>
+        global_value_states;
+
+    inline global_value_state &ensure_global_value_state(const void *obj) {
+        if (!obj) {
+            static thread_local global_value_state fallback;
+            return fallback;
+        }
+        return global_value_states[obj];
+    }
+
+    inline const global_value_state *lookup_global_value_state(const void *obj) {
+        auto it = global_value_states.find(obj);
+        if (it == global_value_states.end())
+            return nullptr;
+        return &it->second;
+    }
+
+    inline void unregister_global_value_state(const void *obj) {
+        if (obj)
+            global_value_states.erase(obj);
+    }
+} // namespace detail
 
 class GlobalValue : public Constant {
 public:
@@ -38,17 +70,43 @@ public:
 
     Type *getValueType() const { return getType(); }
 
-    void setLinkage(LinkageTypes lt) { (void)lt; }
-    LinkageTypes getLinkage() const { return ExternalLinkage; }
+    void setLinkage(LinkageTypes lt) {
+        detail::ensure_global_value_state(this).linkage = (int)lt;
+    }
+    LinkageTypes getLinkage() const {
+        const detail::global_value_state *state =
+            detail::lookup_global_value_state(this);
+        if (!state)
+            return ExternalLinkage;
+        return (LinkageTypes)state->linkage;
+    }
 
-    void setVisibility(VisibilityTypes vt) { (void)vt; }
-    VisibilityTypes getVisibility() const { return DefaultVisibility; }
+    void setVisibility(VisibilityTypes vt) {
+        detail::ensure_global_value_state(this).visibility = (int)vt;
+    }
+    VisibilityTypes getVisibility() const {
+        const detail::global_value_state *state =
+            detail::lookup_global_value_state(this);
+        if (!state)
+            return DefaultVisibility;
+        return (VisibilityTypes)state->visibility;
+    }
 
-    void setUnnamedAddr(UnnamedAddr ua) { (void)ua; }
-    UnnamedAddr getUnnamedAddr() const { return None; }
+    void setUnnamedAddr(UnnamedAddr ua) {
+        detail::ensure_global_value_state(this).unnamed_addr = (int)ua;
+    }
+    UnnamedAddr getUnnamedAddr() const {
+        const detail::global_value_state *state =
+            detail::lookup_global_value_state(this);
+        if (!state)
+            return None;
+        return (UnnamedAddr)state->unnamed_addr;
+    }
 
     bool isDeclaration() const { return false; }
-    bool hasExternalLinkage() const { return true; }
+    bool hasExternalLinkage() const {
+        return getLinkage() == ExternalLinkage;
+    }
 
     void eraseFromParent() {}
 };

--- a/include/llvm/IR/GlobalVariable.h
+++ b/include/llvm/IR/GlobalVariable.h
@@ -26,6 +26,7 @@ public:
     GlobalVariable() = default;
     ~GlobalVariable() {
         detail::unregister_value_wrapper(this);
+        detail::unregister_global_value_state(this);
     }
 
     GlobalVariable(Module &M, Type *Ty, bool isConstant,

--- a/include/llvm/IR/IRBuilder.h
+++ b/include/llvm/IR/IRBuilder.h
@@ -952,17 +952,23 @@ public:
         lr_type_t *arr_ty = lr_type_array_new(lc_module_get_ir(mod), elem_ty,
                                               data.size());
         std::string generated_name;
-        const char *name = nullptr;
+        std::string actual_name;
         std::string explicit_name = Name.str();
         if (!explicit_name.empty()) {
-            name = explicit_name.c_str();
+            actual_name = Module::linkageScopedGlobalName(
+                mod, explicit_name, GlobalValue::PrivateLinkage);
         } else {
             static thread_local unsigned long long str_id = 0;
             generated_name = ".str." + std::to_string(str_id++);
-            name = generated_name.c_str();
+            actual_name = Module::linkageScopedGlobalName(
+                mod, generated_name, GlobalValue::PrivateLinkage);
         }
         lc_value_t *gv = lc_global_create(
-            mod, name, arr_ty, true, data.data(), data.size());
+            mod, actual_name.c_str(), arr_ty, true, data.data(), data.size());
+        if (!explicit_name.empty() && gv && gv->kind == LC_VAL_GLOBAL &&
+            gv->global.name && explicit_name != gv->global.name) {
+            detail::register_global_alias(mod, explicit_name, gv->global.name);
+        }
         return static_cast<Constant *>(Value::wrap(gv));
     }
 

--- a/include/llvm/IR/LLVMContext.h
+++ b/include/llvm/IR/LLVMContext.h
@@ -4,11 +4,13 @@
 #include <liric/liric_compat.h>
 #include <cstdlib>
 #include <cstring>
+#include <string>
 #include <unordered_map>
 
 namespace llvm {
 
 class Function;
+class Module;
 class LLVMContext;
 
 #if defined(__GNUC__) || defined(__clang__)
@@ -36,6 +38,8 @@ namespace detail {
         type_contexts;
     inline thread_local std::unordered_map<const lr_type_t *, vector_type_info>
         vector_types;
+    inline thread_local std::unordered_map<const lc_module_compat_t *,
+        std::unordered_map<std::string, std::string>> global_aliases;
 
     inline void register_value_wrapper(const void *obj, lc_value_t *v) {
         if (obj && v) value_wrappers[obj] = v;
@@ -48,6 +52,32 @@ namespace detail {
 
     inline void unregister_value_wrapper(const void *obj) {
         if (obj) value_wrappers.erase(obj);
+    }
+
+    inline void register_global_alias(const lc_module_compat_t *mod,
+                                      const std::string &logical_name,
+                                      const std::string &actual_name) {
+        if (!mod || logical_name.empty() || actual_name.empty())
+            return;
+        global_aliases[mod][logical_name] = actual_name;
+    }
+
+    inline std::string lookup_global_alias(const lc_module_compat_t *mod,
+                                           const std::string &logical_name) {
+        if (!mod || logical_name.empty())
+            return std::string();
+        auto mit = global_aliases.find(mod);
+        if (mit == global_aliases.end())
+            return std::string();
+        auto ait = mit->second.find(logical_name);
+        if (ait == mit->second.end())
+            return std::string();
+        return ait->second;
+    }
+
+    inline void clear_global_aliases(const lc_module_compat_t *mod) {
+        if (mod)
+            global_aliases.erase(mod);
     }
 
     inline void register_function_wrapper(const lr_func_t *f, Function *fn) {

--- a/include/llvm/IR/Module.h
+++ b/include/llvm/IR/Module.h
@@ -17,6 +17,8 @@
 #include <string>
 #include <vector>
 #include <memory>
+#include <cstdint>
+#include <cstdio>
 
 namespace llvm {
 
@@ -57,6 +59,7 @@ public:
         if (current_ == compat_) current_ = nullptr;
         if (detail::fallback_module == compat_)
             detail::fallback_module = ctx_.getDefaultModule();
+        detail::clear_global_aliases(compat_);
         lc_module_destroy(compat_);
     }
 
@@ -69,6 +72,25 @@ public:
         return LLVMContext::getGlobal().getDefaultModule();
     }
     static void setCurrentModule(lc_module_compat_t *m) { current_ = m; }
+
+    static bool isLocalGlobalLinkage(GlobalValue::LinkageTypes linkage) {
+        return linkage == GlobalValue::InternalLinkage ||
+               linkage == GlobalValue::PrivateLinkage;
+    }
+
+    static std::string linkageScopedGlobalName(
+        lc_module_compat_t *compat, StringRef name,
+        GlobalValue::LinkageTypes linkage) {
+        std::string base = name.str();
+        if (!compat || base.empty() || !isLocalGlobalLinkage(linkage))
+            return base;
+        char suffix[32];
+        std::snprintf(suffix, sizeof(suffix), "%llx",
+                      (unsigned long long)(uintptr_t)compat);
+        base += ".__liric_local.";
+        base += suffix;
+        return base;
+    }
 
     lc_module_compat_t *getCompat() const { return compat_; }
     lr_module_t *getIR() const { return lc_module_get_ir(compat_); }
@@ -114,6 +136,9 @@ public:
         (void)AllowInternal;
         Module::setCurrentModule(compat_);
         std::string global_name = Name.str();
+        std::string alias_name = detail::lookup_global_alias(compat_, global_name);
+        if (!alias_name.empty())
+            global_name = alias_name;
 
         for (const auto &og : owned_globals_) {
             lc_value_t *gv = detail::lookup_value_wrapper(og.get());
@@ -198,19 +223,33 @@ public:
         const char *name, Type *ty, bool is_const,
         GlobalValue::LinkageTypes linkage,
         const void *init_data = nullptr, size_t init_size = 0) {
-        (void)linkage;
         Module::setCurrentModule(compat_);
+        std::string requested_name = name ? name : "";
+        std::string actual_name =
+            linkageScopedGlobalName(compat_, requested_name, linkage);
+        const char *symbol_name = actual_name.empty() ? name : actual_name.c_str();
+        if (!symbol_name)
+            symbol_name = "";
         lc_value_t *gv;
-        if (init_data && init_size > 0) {
-            gv = lc_global_create(compat_, name, ty->impl(), is_const,
+        bool must_define = (init_data && init_size > 0) ||
+                           isLocalGlobalLinkage(linkage);
+        if (must_define) {
+            gv = lc_global_create(compat_, symbol_name, ty->impl(), is_const,
                                   init_data, init_size);
         } else {
-            gv = lc_global_declare(compat_, name, ty->impl());
+            gv = lc_global_declare(compat_, symbol_name, ty->impl());
         }
         auto g = std::make_unique<GlobalVariable>();
         (void)gv;
         GlobalVariable *ptr = g.get();
         detail::register_value_wrapper(ptr, gv);
+        ptr->setLinkage(linkage);
+        std::string final_name = actual_name;
+        if (gv && gv->kind == LC_VAL_GLOBAL && gv->global.name)
+            final_name = gv->global.name;
+        if (!requested_name.empty() && requested_name != final_name) {
+            detail::register_global_alias(compat_, requested_name, final_name);
+        }
         owned_globals_.push_back(std::move(g));
         return ptr;
     }
@@ -435,6 +474,7 @@ inline GlobalVariable::GlobalVariable(Module &M, Type *Ty, bool isConstant,
     if (created) {
         detail::register_value_wrapper(this,
             detail::lookup_value_wrapper(created));
+        setLinkage(Linkage);
         if (Initializer) {
             Module::setCurrentModule(M.getCompat());
             setInitializer(Initializer);

--- a/tests/test_llvm_compat.cpp
+++ b/tests/test_llvm_compat.cpp
@@ -519,6 +519,44 @@ static int test_duplicate_global_names_are_uniquified() {
     return 0;
 }
 
+static int test_private_global_strings_are_module_scoped() {
+    llvm::LLVMContext ctx;
+    llvm::Module mod_a("private_strings_a", ctx);
+    llvm::Module mod_b("private_strings_b", ctx);
+
+    llvm::IRBuilder<> builder_a(ctx);
+    builder_a.SetModule(mod_a.getCompat());
+    llvm::IRBuilder<> builder_b(ctx);
+    builder_b.SetModule(mod_b.getCompat());
+
+    llvm::Constant *a_named = builder_a.CreateGlobalStringPtr("ERROR STOP", "ERROR STOP");
+    llvm::Constant *b_named = builder_b.CreateGlobalStringPtr("ERROR STOP", "ERROR STOP");
+    TEST_ASSERT(a_named != nullptr, "named string created in first module");
+    TEST_ASSERT(b_named != nullptr, "named string created in second module");
+    TEST_ASSERT(a_named->impl()->kind == LC_VAL_GLOBAL, "first named string is global");
+    TEST_ASSERT(b_named->impl()->kind == LC_VAL_GLOBAL, "second named string is global");
+    TEST_ASSERT(a_named->impl()->global.name != nullptr, "first named global has name");
+    TEST_ASSERT(b_named->impl()->global.name != nullptr, "second named global has name");
+    TEST_ASSERT(std::strcmp(a_named->impl()->global.name,
+                            b_named->impl()->global.name) != 0,
+                "named private global symbols are module-scoped");
+    TEST_ASSERT(mod_a.getNamedGlobal("ERROR STOP") != nullptr,
+                "first module resolves named private global by original name");
+    TEST_ASSERT(mod_b.getNamedGlobal("ERROR STOP") != nullptr,
+                "second module resolves named private global by original name");
+
+    llvm::Constant *a_auto = builder_a.CreateGlobalStringPtr("x");
+    llvm::Constant *b_auto = builder_b.CreateGlobalStringPtr("x");
+    TEST_ASSERT(a_auto != nullptr, "auto string created in first module");
+    TEST_ASSERT(b_auto != nullptr, "auto string created in second module");
+    TEST_ASSERT(a_auto->impl()->global.name != nullptr, "first auto global has name");
+    TEST_ASSERT(b_auto->impl()->global.name != nullptr, "second auto global has name");
+    TEST_ASSERT(std::strcmp(a_auto->impl()->global.name,
+                            b_auto->impl()->global.name) != 0,
+                "auto private global symbols are module-scoped");
+    return 0;
+}
+
 static int test_function_creation() {
     llvm::LLVMContext ctx;
     llvm::Module mod("funcs", ctx);
@@ -667,8 +705,11 @@ static int test_builder_syncs_module_from_insert_block() {
     lr_module_t *m = lc_module_get_ir(mod.getCompat());
     TEST_ASSERT(m != nullptr, "module ir handle");
     TEST_ASSERT(m->first_global != nullptr, "global was created");
-    TEST_ASSERT(std::strcmp(m->first_global->name, "serialization_info") == 0,
-                "global has correct name");
+    TEST_ASSERT(m->first_global->name != nullptr, "global has symbol name");
+    TEST_ASSERT(std::strstr(m->first_global->name, "serialization_info") != nullptr,
+                "global symbol keeps original name as prefix");
+    TEST_ASSERT(mod.getNamedGlobal("serialization_info") != nullptr,
+                "named global lookup resolves original string name");
     TEST_ASSERT(main_fn->getIRFunc() != nullptr, "main IR func exists");
     return 0;
 }
@@ -1780,6 +1821,7 @@ int main() {
     RUN_TEST(test_global_lookup_set_initializer_and_jit);
     RUN_TEST(test_create_global_without_initializer_is_declaration);
     RUN_TEST(test_duplicate_global_names_are_uniquified);
+    RUN_TEST(test_private_global_strings_are_module_scoped);
     RUN_TEST(test_parse_assembly_wrapper_fast_path);
 
     fprintf(stderr, "\nFunction tests:\n");

